sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala

```diff
@@ -126,4 +136,52 @@ object FileSourceStrategy extends Strategy with Logging {
 
     case _ => Nil
   }
+
+  private def generateStructFieldsContainsNesting(projects: Seq[Expression],
+                                      totalSchema: StructType) : Seq[StructField] = {
+    def generateStructField(curField: List[String],
+                             node: Expression) : Seq[StructField] = {
+      node match {
+        case ai: GetArrayItem =>
+          // Here we drop the previous for simplify array and map support.
+          // Same strategy in GetArrayStructFields and GetMapValue
+          generateStructField(List.empty[String], ai.child)
+        case asf: GetArrayStructFields =>
+          generateStructField(List.empty[String], asf.child)
+        case mv: GetMapValue =>
+          generateStructField(List.empty[String], mv.child)
+        case attr: AttributeReference =>
+          Seq(getFieldRecursively(totalSchema, attr.name :: curField))
+        case sf: GetStructField =>
+          generateStructField(sf.name.get :: curField, sf.child)
```

Comment:
fix done.
a little question here, all projects parse from sql must have the name while projects from dataframe api may not   , right?